main.cpp

```diff
@@ -568,157 +769,247 @@ bool is_valid_ip(udp::endpoint const& ep)
+		memcpy(remote_ip, ip.data(), ip.size());
+		remote_ip[ip.size()] = (ep.port() >> 8) & 0xff;
+		remote_ip[ip.size() + 1] = ep.port() & 0xff;
+		return ip.size() + 2;
+	}
+}
+
+std::array<char, 4> version = {{0, 0, 0, 0}};
 
-	for (;;)
+struct router_thread
+{
+	router_thread(std::vector<address> addrs)
+		: signals(ios)
 	{
-		udp::endpoint ep;
-		error_code ec;
+		for (address a : addrs)
+		{
+			socks.emplace_back(ios, udp::endpoint(a, 6881));
+		}
 
-#ifdef DEBUG_STATS
-		nodebuf_size[threadid] = node_buffer.size();
-#endif
+		signals.add(SIGINT);
+		signals.add(SIGTERM);
+
+		// close the sockets and stop the io service when signalled to quit
+		signals.async_wait(std::bind(&router_thread::signal_handler, this, _1, _2));
+	}
 
-		// rotate the secrets every 10 minutes
-		steady_clock::time_point now = steady_clock::now();
-		if (last_secret_rotate + minutes(10) < now)
+	void signal_handler(error_code const& e, int /*signo*/)
+	{
+		for (bound_socket& sock : socks)
 		{
-			std::lock_guard<std::mutex> l(secret_mutex);
-			if (last_secret_rotate + minutes(10) < now)
-			{
-				last_secret_rotate = now;
-
-				rand.seed(r());
-				// there are three buffers in total and two of
-				// them are exposed via secret1 and secret2 at
-				// any given time. The old_secret needs to be
-				// kept around for a bit to avoid race conditions
-				// with threads that may have just grabbed the
-				// pointer to it. It's rotated into the intermediate
-				// slot.
-				uint8_t* old_secret = secret2;
-				secret2 = secret1.load();
-
-				uint8_t* i = intermediate.load();
-				std::generate(i, i + 20, [&](){ return random_byte(rand);});
-				secret1 = i;
-				intermediate = old_secret;
-			}
+			error_code ec;
+			sock.sock.close(ec);
+			if (ec)
+				fprintf(stderr, "socket: (%d) %s\n"
+					, ec.value(), ec.message().c_str());
 		}
+		ios.stop();
+	}
 
-		// if we need to ping nodes, do that now
-		queued_node_t n;
-		while (ping_queue.need_ping(&n))
+	void start(int threadid)
+	{
+		if (socks.empty())
 		{
-//			fprintf(stderr, "pinging node\n");
-			// build the IP field
-			char remote_ip[6];
-			address_v4::bytes_type ip = n.ep.address().to_v4().to_bytes();
-			memcpy(remote_ip, &ip[0], 4);
-			remote_ip[4] = (n.ep.port() >> 8) & 0xff;
-			remote_ip[5] = n.ep.port() & 0xff;
-
-			// compute transaction ID
-			std::string transaction_id = compute_tid(secret1, remote_ip);
-
-			// send the ping to this node
-			bencoder b(response, sizeof(response));
-			b.open_dict();
+			fprintf(stderr, "no interfaces to receive on\n");
+			return;
+		}
 
-			b.add_string("ip"); b.add_string(remote_ip, 6);
+		this->threadid = threadid;
+		printf("starting thread %d\n", threadid);
 
-			// args dict
-			b.add_string("a");
-			b.open_dict();
-			b.add_string("id"); b.add_string(our_node_id, 20);
-			b.close_dict();
+		for (bound_socket& sock : socks)
+		{
+			sock.sock.async_receive_from(buffer(packet, sizeof(packet)), ep
```

Comment:
for multiple sockets, receive data in same buffer? I guess even in same thread, the buffer maybe overwritten before/when you handle it.

I have some codes like this, which definitely wrong, I guess it's similar to this code.
```c++
void on_packet_received(error_code const& ec, size_t len)
{
    // actually should read data now, but ...

     sock.async_receive_from(same-buffer, endpoint, on_packet_received);

    // read data from same-buffer, which in some cases, it's already overwritten by next packet.
}